File: lightroom_tagger/core/score_perspective.py
"""Validate ``score_perspective`` values by catalog row existence."""

from __future__ import annotations

import sqlite3

from lightroom_tagger.core.database import get_perspective_by_slug


def validate_score_perspective_exists(
    db: sqlite3.Connection,
    slug: str | None,
) -> tuple[str | None, str | None]:
    """Return ``(normalized_slug, error_message)``.

    Empty/absent slug → ``(None, None)``. Unknown slug → ``(None, "unknown perspective '…'")``.
    Existence is satisfied by any row in ``perspectives``, regardless of ``active``.
    """
    sp = (slug or "").strip() or None
    if sp is None:
        return None, None
    if get_perspective_by_slug(db, sp) is None:
        return None, f"unknown perspective '{sp}'"
    return sp, None

File: apps/visualizer/backend/src/utils/score-perspective.ts
/**
 * Validate `score_perspective` values by catalog row existence.
 */
import type { Db } from '../db/connection.js';
import { getPerspectiveBySlug } from '../db/library/scores.js';

/**
 * Resolve a `score_perspective` query value.
 *
 * An empty or absent slug is not an error — it means "no perspective filter".
 * Existence is satisfied by any row in `perspectives` regardless of `active`, so a
 * deactivated perspective can still be used to sort or filter historical scores.
 */
export function validateScorePerspectiveExists(
  db: Db,
  slug: string | null | undefined,
): { slug: string | null; error: string | null } {
  const sp = (slug ?? '').trim() || null;
  if (sp === null) return { slug: null, error: null };
  if (getPerspectiveBySlug(db, sp) === null) {
    return { slug: null, error: `unknown perspective '${sp}'` };
  }
  return { slug: sp, error: null };
}
